### PR TITLE
[WIP] JS refactor->rename for named variables, members and functions in a single file

### PR DESCRIPTION
**Pull request by swmitra**
This PR implements "rename" functionality in JS mode to enable intelligent rename feature by using Terns scope analysis and inference functions. Renaming across multiple files has been intentionally blocked for the time being as we are not analyzing all the JS files in current project root right now. Once the other PR #11948 is merged, we can leverage on the JS analysis in node and include all files to build symbol table. 
The rename function works by selecting a variable def/ref or placing cursor in and using Ctrl+R. 

Know issues
--------------
In a particular case Tern inference engine is picking up wrong hit. For example - 
 `function fn1(error) {` 
   `     error.a = "abc";`
   `     var a = "test";`
   `     console.log(a);`
 `}`
If we try to rename `a` in line 3 or 4 , Tern picks up fake ref from line 2 as well. 

Yet to be done
----------------

- [x] To enable context menu option for rename
- [ ] Update JS doc and minor code cleanup  
- [ ] Couple of bug fixes ( one mentioned above and another to center primary selection )

### CODE REVIEW

File: src/extensions/default/JavaScriptCodeHints/main.js [hunk 1]
```diff
@@ -859,6 +859,62 @@ define(function (require, exports, module) {
...
         /*
+         * Handle Rename menu/keyboard command.
+         */
+        function handleRename() {
+            var offset,
+                handleFindRefs;
+
+
+            // Only provide rename when cursor is in JavaScript content
+            if (!session || session.editor.getModeForSelection() !== "javascript") {
```
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/main.js]: Can `session.editor` be `null` or `undefined`? Most likely not, but might be worth checking out.

File: src/extensions/default/JavaScriptCodeHints/ScopeManager.js [hunk 2]
```diff
@@ -392,6 +392,40 @@ define(function (require, exports, module) {
...
+            fileInfo: fileInfo,
+            offset: offset
+        });
+
+        return addPendingRequest(fileInfo.name, offset, MessageIds.TERN_REFS);
+    }
+    
+    function requestFindRefs(session, document, offset) {
+        var path    = document.file.fullPath,
+            fileInfo = {type: MessageIds.TERN_FILE_INFO_TYPE_FULL,
```
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/ScopeManager.js]: Nit: Have we used some kind of style for multiline objects? The indentation looks a bit weird, I would prefer
```js
fileInfo = {
  type: MessageIds.TERN_FILE_INFO_TYPE_FULL,
  name: path,
  offsetLines: 0,
  text: filterText(session.getJavascriptText())
};

File: src/extensions/default/JavaScriptCodeHints/main.js [hunk 3]
```diff
@@ -859,6 +859,62 @@ define(function (require, exports, module) {
...
+
+            /**
+             * Make a find ref request.
+             * @param {Session} session - the session
+             * @param {number} offset - the offset of where to jump from
+             */
+            function requestFindRefs(session, offset) {
+                var response = ScopeManager.requestFindRefs(session, session.editor.document, offset);
+
+                if (response.hasOwnProperty("promise")) {
```
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/main.js]: Is there are case where response wouldn't have a `promise`? `requestFindRefs` looks like it's pretty null safe, but might be good to guard with `response && response.hasOwn...`

File: src/extensions/default/JavaScriptCodeHints/main.js [hunk 4]
```diff
@@ -859,6 +859,62 @@ define(function (require, exports, module) {
...
+                    response.promise.done(handleFindRefs).fail(function () {
```
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/main.js]: Can we use `then` and `catch` to align the old Deferred syntax with the more modern Promises/A+ syntax?

File: src/extensions/default/JavaScriptCodeHints/main.js [hunk 5]
```diff
@@ -859,6 +859,62 @@ define(function (require, exports, module) {
...
+                }
+            }
+
+            /**
+             * handle processing of the completed jump-to-def request.
+             * will open the appropriate file, and set the selection based
+             * on the response.
+             */
+            handleFindRefs = function (refsResp) {
+                function isInSameFile(obj) {
```
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/main.js]: Can be shortened to `return obj && obj.file === refsResp.file`

File: src/extensions/default/JavaScriptCodeHints/main.js [hunk 6]
```diff
@@ -859,6 +859,62 @@ define(function (require, exports, module) {
...
+                    if (obj && obj.file === refsResp.file) {
+                        return true;
+                    } else {
+                        return false;
+                    }
+                }
+                
+                if (refsResp && refsResp.refs && refsResp.refs.refs) {
```
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/main.js]: Random comment: `refsResp.refs.refs` looks kinda silly 👯

File: src/extensions/default/JavaScriptCodeHints/tern-worker.js [hunk 7]
```diff
@@ -231,6 +231,27 @@ var config = {};
...
+    
+            function getRefs(fileInfo, offset) {
+                var request = buildRequest(fileInfo, "refs", offset);
+                try {
+                    ternServer.request(request, function (error, data) {
+                        if (error) {
+                            _log("Error returned from Tern 'definition' request: " + error);
+                            return;
+                        }
+                        var response = {type: MessageIds.TERN_REFS,
```
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/tern-worker.js]: NIT: Same thing here with the new lines and objects

**Review by petetnt — CHANGES_REQUESTED** [review]
Some initial comments here. Great work so far!

**Comment by swmitra** [comment]
Thanks a lot @petetnt for the quick review. I will try to address the comments asap. I would urge all of you to use the feature as well and let me know if it's helpful.

File: src/base-config/keyboard.json [hunk 8]
```diff
@@ -329,5 +329,8 @@
     ],
     "help.support": [
         "F1"
+    ],
+    "refactor.rename": [
+        "Ctrl-R"
```
Comment (ficristo) [inline src/base-config/keyboard.json]: I believe on macOS this is translated to Cmd+R which is used to reload Brackets with Extensions.

File: src/extensions/default/JavaScriptCodeHints/main.js [hunk 9]
```diff
@@ -867,6 +923,8 @@ define(function (require, exports, module) {
 
             return response;
         }
+        
+        CommandManager.register("Rename", "refactor.rename", handleRename);
```
Comment (ficristo) [inline src/extensions/default/JavaScriptCodeHints/main.js]: You should add the string in `command/Commands.js` instead of using the string explicit.

**Review by ficristo — CHANGES_REQUESTED** [review]

**Comment by ficristo** [comment]
From the code, this seems to work only for javascript files. I assume languages that derive from javascript in CodeMirror are NOT supported (for example json, typescript, ...)
Is it possible for extensions to easily override this behaviour so to support other languages?

**Comment by swmitra** [comment]
@ficristo Yes, it's targeted for javascript mode.  So if CM identifies any block of code as javascript,  this would work there.  That includes script blocks in html,  any modes derived from JS by CodeMirror.  BTW,    is there any extension available today which implements this feature?

**Comment by swmitra** [comment]
@ficristo Yes, it's targeted for javascript mode.  So if CM identifies any block of code as javascript,  this would work there.  That includes script blocks in html,  any modes derived from JS by CodeMirror.  BTW,    is there any extension available today which implements this feature?

**Comment by abose** [comment]
@swmitra  Can you post a gif of how this works?

**Comment by swmitra** [comment]
@abose I am not able to post a gif right now as I dont have access to any computer for a few days.  But what you can do is - put the cursor on top of the token you want to rename.  Then press ctrl+r . This will put selection on all the occurrences after doing a scope analysis.  Now just type the new name.  You will see multi cursor in action and should be able to rename all the occurrences.

**Comment by abose** [comment]
👍 Cool feature

**Comment by ficristo** [comment]
> is there any extension available today which implements this feature?

Looking at #11658 maybe [ternific](https://github.com/MiguelCastillo/Brackets-Ternific)?

**Pull request closed without merging**